### PR TITLE
Introduce GEOPM LDMS Sampler

### PR DESCRIPTION
**Pull request by cmcantalupo**
- Create new sampler plugin
- Started with meminfo plugin as a template
- Add optional support to link libgeopmd into the build
- Enable configuration of GEOPM sampler through file input
- Support for comments and blank lines in configuration file
- Plugin library factored with non-static functions as test entry points
- The ldms_geopm_sampler.h file defines test entry points
- One test binary is called by several test scripts
- Uses the make check target from autotools
- Add third-party-programs file for GEOPM Sampler
- Add test for valid configuration file
- Add test for file with extra data on line
- Add test for truncated configuration file
- Add test for bad domain name
- Add test for out of range domain index
- Add test for invalid signal
- Add test for missing config file

Co-authored-by: Alejandro Vilches <alejandro.vilches@intel.com>
Co-authored-by: Siddhartha Jana <siddhartha.jana@intel.com>
Co-authored-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>
Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

### CODE REVIEW

**Comment by cmcantalupo** [comment]
Hi @baallan,

Within the PR is a README.md where there are some links and build information information about the GEOPM plugin:

https://github.com/cmcantalupo/ovis/tree/geopm-sampler/ldms/src/contrib/sampler/geopm_sampler#geopm-ldms-plugin

Please let me know if I can answer any questions or comments about GEOPM or the sampler.  Would it be possible to enable the github workflows for this PR?

Thanks!
Chris

File: ldms/src/contrib/sampler/geopm_sampler/ldms_geopm_sampler.c [hunk 1]
```diff
@@ -0,0 +1,573 @@
...
+				   "%%%ds %%%ds %%d %%%ds ",
+				   NAME_MAX - 1, NAME_MAX - 1, NAME_MAX - 1);
+	if (format_size >= NAME_MAX) {
+		return ENAMETOOLONG;
+	}
+
+	char domain_type_str[NAME_MAX];
+	char *line = NULL;
+	size_t line_len = 0;
+	int num_scan = 0;
```
Comment (baallan) [inline ldms/src/contrib/sampler/geopm_sampler/ldms_geopm_sampler.c:185]: set errno = 0; before depending on it as a check.

**Comment by cmcantalupo** [comment]
Thanks for the feedback @baallan and @morrone!  I'll address these issues shortly.

**Comment by baallan** [comment]
@cmcantalupo if the libraries the new sampler depends on are not regularly tested with valgrind or equivalent, please test ldmsd+sampler with valgrind for memory use errors also, particularly in the odd cases of a failed configuration.

File: ldms/src/contrib/sampler/geopm_sampler/ldms_geopm_sampler.c (hunk 1 above)
Comment (avilcheslopez) [reply ldms/src/contrib/sampler/geopm_sampler/ldms_geopm_sampler.c:184]: Issue addressed. Let us know if it looks good now.

File: ldms/src/contrib/sampler/geopm_sampler/ldms_geopm_sampler.c [hunk 2]
```diff
@@ -0,0 +1,573 @@
...
+					  domain_type_str, domain_idx, unparsed);
+			if (num_scan >= 1 && signal_name[0] == '#') {
+				num_scan = 0;
+			}
+		}
+	}
+	if (line != NULL) {
+		free(line);
+	}
+	if (num_scan == EOF) {
```
[the remark this answers is not in the document]
Comment (avilcheslopez) [reply ldms/src/contrib/sampler/geopm_sampler/ldms_geopm_sampler.c:215]: Addressed.

File: ldms/src/contrib/sampler/geopm_sampler/ldms_geopm_sampler.c [hunk 3]
```diff
@@ -0,0 +1,573 @@
...
+
+static void term(struct ldmsd_plugin *self)
+{
+	if (g_base) {
+		base_del(g_base);
+	}
+	if (g_set) {
+		ldms_set_delete(g_set);
+	}
+	g_set = NULL;
```
[the remark this answers is not in the document]
Comment (avilcheslopez) [reply ldms/src/contrib/sampler/geopm_sampler/ldms_geopm_sampler.c:562]: Currently looking into this...

**Review by avilcheslopez — COMMENTED** [review]
@baallan: thanks for the review feedback. We've addressed some of the comments and we're currently looking into addressing the comment about memory clean-up/Valgrind.

File: ldms/src/contrib/sampler/geopm_sampler/ldms_geopm_sampler.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (cmcantalupo) [reply ldms/src/contrib/sampler/geopm_sampler/ldms_geopm_sampler.c:562]: I added the option to run the unit tests under any analysis wrapper here:

https://github.com/ovis-hpc/ovis/pull/955/commits/00a5ceb24a4fec790436ca06056c43dd6d33be9a

and got a clean report from valgrind.

**Comment by cmcantalupo** [comment]
I added the option to run the unit tests under any analysis tool wrapper here:

https://github.com/ovis-hpc/ovis/pull/955/commits/00a5ceb24a4fec790436ca06056c43dd6d33be9a

and got a clean report from valgrind.  Note that we have negative tests for all the ways that a configuration file may be mal-formatted.  These tests check the output to the error log and the value of the return code for these error conditions.  When the tests are run under valgrind this enables memory checking during the error condition.

File: ldms/src/contrib/sampler/geopm_sampler/ldms_geopm_sampler.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (cmcantalupo) [reply ldms/src/contrib/sampler/geopm_sampler/ldms_geopm_sampler.c:562]: This points to a gap in our API's for this use case.  After the batch of signals is read once, PlatformIO's batch of signals cannot be reconfigured.  There is a simple solution for this, and we are adding it to our PlatformIO APIs.  This is a work in progress, but we will include this geopm_pio_reset() API as part of our next release candidate (geopm v2.0.0+rc2). Once we have published a release that contains this API we will update the sampler.

The WIP PR is posted here:
https://github.com/geopm/geopm/pull/2404/

File: ldms/src/contrib/sampler/geopm_sampler/Plugin_geopm_sampler.man [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+name=<plugin_name>
+.br
+This MUST be geopm.
+.TP
+schema=<schema>
+.br
+Optional schema name. It is intended that the same sampler on
+different nodes with different metrics have a different schema. If not
+specified, will default to `geopm`.
+.RE
```
[the remark this answers is not in the document]
Comment (sidjana) [reply ldms/src/contrib/sampler/geopm_sampler/Plugin_geopm_sampler.man:65]: Addressed.

[the remark this answers is not in the document]
Comment (avilcheslopez) [reply ldms/src/contrib/sampler/geopm_sampler/Plugin_geopm_sampler.man:69]: Addressed. Please let us know if you have any further feedback.

File: ldms/src/contrib/sampler/geopm_sampler/Plugin_geopm_sampler.man [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+     -------------- ------------------------ ------ ------ ------ ------ ------ ------ ---------- ----------------- ----------------- --------
+     Total Sets: 1, Meta Data (kB): 1.35, Data (kB) 0.24, Memory (kB): 1.59
+     
+     =======================================================================
+     
+     <hostname>/ldms_geopm_sampler: consistent, last update: Tue Jun 28 08:46:33 2022 -0700 [51578us]
+     M u64        component_id                               1
+     D u64        job_id                                     0
+     D u64        app_id                                     0
+     D d64        CPU_FREQUENCY_MAX-board-0                  3700000000.000000
```
Comment (tom95858) [inline ldms/src/contrib/sampler/geopm_sampler/Plugin_geopm_sampler.man:148]: Hi Guys,

Although '-' is present in a lot of other schema, could we replace them with '_'? The reason is that if we store these items in the database and then try to query it with a NoSQL syntax, the '-' is a special character that we might want to use in the future for server side operations.

We're trying to get away from these special characters in metric names. Specifically, '(', ')', '-', '+', etc...

File: ldms/src/contrib/sampler/geopm_sampler/README.md [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+
+
+Unit Tests
+----------
+
+The GEOPM LDMS sampler unit tests are executed as part of the
+``make check`` target of the OVIS autotools build system.  The test
+binary ``ldms_geopm_sampler_test`` can be used to test the validity of
+any GEOPM LDMS sampler configuration file.  There are scripts that
+execute this binary on a a set of configuration files, and then
```
Comment (cmcantalupo) [inline ldms/src/contrib/sampler/geopm_sampler/README.md:83]:
```suggestion
execute this binary on a set of configuration files, and then
```

**Comment by cmcantalupo** [comment]
@tom95858 

Rebased against OVIS-4 and conflict resolved.

**Comment by cmcantalupo** [comment]
Thanks @tom95858 for merging our sampler!  We have tagged a new release candidate of GEOPM.  This RC provides the ``geopm_pio_reset()`` API that would resolve the [restart problem](https://github.com/ovis-hpc/ovis/pull/955#discussion_r902927182).  I will create a new PR soon that adds a call to ``geopm_pio_reset()`` in the term callback and updates the documentation to refer to the newer GEOPM release candidate.
